crates/client/db/src/bonsai_db.rs

```diff
@@ -1,265 +1,404 @@
-        _batch: Option<&mut Self::Batch>,
+        key: &DatabaseKey,
+        batch: Option<&mut Self::Batch>,
     ) -> Result<Option<Vec<u8>>, Self::DatabaseError> {
-        Ok(None)
+        log::trace!("Removing from RocksDB: {:?}", key);
+        let handle = self.db.get_column(self.column_mapping.map(key));
+        let old_value = self.txn.get_cf(&handle, key.as_slice())?;
+        if let Some(batch) = batch {
+            batch.delete_cf(&handle, key.as_slice());
+        } else {
+            self.txn.delete_cf(&handle, key.as_slice())?;
+        }
+        Ok(old_value)
     }
 
-    /// Removes all key-value pairs starting with a given prefix.
-    fn remove_by_prefix(&mut self, _prefix: &DatabaseKey) -> Result<(), Self::DatabaseError> {
+    fn remove_by_prefix(&mut self, prefix: &DatabaseKey) -> Result<(), Self::DatabaseError> {
+        log::trace!("Getting from RocksDB: {:?}", prefix);
+        let handle = self.db.get_column(self.column_mapping.map(prefix));
+        let iter = self.txn.iterator_cf(&handle, IteratorMode::From(prefix.as_slice(), Direction::Forward));
+        let mut batch = self.create_batch();
+        for kv in iter {
+            if let Ok((key, _)) = kv {
+                if key.starts_with(prefix.as_slice()) {
+                    batch.delete_cf(&handle, &key);
+                } else {
+                    break;
+                }
+            } else {
+                break;
+            }
+        }
+        drop(handle);
+        self.write_batch(batch)?;
         Ok(())
     }
 
-    /// Writes a batch of changes to the database.
-    fn write_batch(&mut self, _batch: Self::Batch) -> Result<(), Self::DatabaseError> {
-        Ok(())
+    fn write_batch(&mut self, batch: Self::Batch) -> Result<(), Self::DatabaseError> {
+        Ok(self.txn.rebuild_from_writebatch(&batch)?)
     }
 }
 
-/// This implementation is a stub to mute any error but is is currently not used.
-impl<ID: Id> BonsaiPersistentDatabase<ID> for BonsaiDb {
-    type Transaction = TransactionWrapper;
+impl<'db> BonsaiPersistentDatabase<BasicId> for BonsaiDb<'db>
+where
+    Self: 'db,
+{
+    type Transaction = BonsaiTransaction<'db>;
     type DatabaseError = BonsaiDbError;
 
-    /// Snapshots the current database state, associated with an ID.
-    fn snapshot(&mut self, _id: ID) {}
+    fn snapshot(&mut self, id: BasicId) {
+        log::trace!("Generating RocksDB snapshot");
+        let snapshot = self.db.snapshot();
+        self.snapshots.insert(id, snapshot);
+        // TODO: snapshot limits
+        // if let Some(max_number_snapshot) = self.config.max_saved_snapshots {
+        //     while self.snapshots.len() > max_number_snapshot {
+        //         self.snapshots.pop_first();
+        //     }
+        // }
+    }
+
+    fn transaction(&self, id: BasicId) -> Option<Self::Transaction> {
+        log::trace!("Generating RocksDB transaction");
+        if let Some(snapshot) = self.snapshots.get(&id) {
```

Comment:
If I got this right, this will only be able to retrieve database transactions if we have stored the _exact_ snapshot for the given `id`?

Reply:
yes, this is taken from https://github.com/keep-starknet-strange/bonsai-trie/blob/oss/src/databases/rocks_db.rs
I don't fully understand snapshots, but from what I understand they have a lifetime that must be less than the DB lifetime: they don't persist across restarts